compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ArrayEqualsOp.java

```diff
@@ -236,4 +272,105 @@ private void emitTailCompares(AArch64MacroAssembler masm, Register result, Regis
+
+        Label compareByChunkHead = new Label();
+        Label compareByChunkTail = new Label();
+        Label processTail = new Label();
+
+        /* 1. Set 'array1Address' and 'array2Address' to point to start of arrays. */
+        loadArrayStart(masm, array1Address, array2Address);
+        /*
+         * 2.1 Set endOfArray1 pointing to byte next to the last valid element in array1 and
+         * 'refAddress1' pointing to the beginning of the last chunk.
+         */
+        masm.add(64, endOfArray1, array1Address, byteArrayLength);
+        masm.sub(64, refAddress1, endOfArray1, 32);
+
+        masm.align(16);
+        masm.bind(compareByChunkHead);
+        masm.cmp(64, refAddress1, array1Address);
+        masm.branchConditionally(ConditionFlag.LS, processTail);
+        masm.bind(compareByChunkTail);
+
+        /* 2.2 Read a 32-byte chunk from source and target array each in two SIMD registers. */
+        masm.fldp(128, array1Part1RegV, array1Part2RegV, AArch64Address.createImmediateAddress(128, AArch64Address.AddressingMode.IMMEDIATE_PAIR_POST_INDEXED, array1Address, 32));
+        masm.fldp(128, array2Part1RegV, array2Part2RegV, AArch64Address.createImmediateAddress(128, AArch64Address.AddressingMode.IMMEDIATE_PAIR_POST_INDEXED, array2Address, 32));
+        /* 3. Compare arrays in the 32-byte chunk */
+        masm.neon.cmeqVVV(AArch64ASIMDAssembler.ASIMDSize.FullReg, AArch64ASIMDAssembler.ElementSize.DoubleWord, array1Part1RegV, array1Part1RegV, array2Part1RegV);
+        masm.neon.cmeqVVV(AArch64ASIMDAssembler.ASIMDSize.FullReg, AArch64ASIMDAssembler.ElementSize.DoubleWord, array1Part2RegV, array1Part2RegV, array2Part2RegV);
+        /* 4. Determining if they are identical. */
+        /* 4.1 Combine two registers into 1 register */
+        masm.neon.andVVV(AArch64ASIMDAssembler.ASIMDSize.FullReg, array1Part1RegV, array1Part1RegV, array1Part2RegV);
+        /* 4.2 Find the minimum value across the vector */
+        masm.neon.uminvSV(AArch64ASIMDAssembler.ASIMDSize.FullReg, AArch64ASIMDAssembler.ElementSize.Word, array1Part1RegV, array1Part1RegV);
+        /* 4.3 If hasMismatch != 0, then there is a mismatch somewhere. */
+        masm.neon.moveFromIndex(AArch64ASIMDAssembler.ElementSize.DoubleWord, AArch64ASIMDAssembler.ElementSize.Word, hasMismatch, array1Part1RegV, 0);
+        masm.add(64, hasMismatch, hasMismatch, 1);
+        /* If there is mismatch, then no more searching is necessary */
+        masm.cbnz(64, hasMismatch, endLabel);
+
+        /* 5. No mismatch; jump to next loop iteration. */
+        // align array1Address to 32-byte boundary
+        // determine how much to subtract from array2Address to match aligned array1Address
+        Register array1Alignment = asRegister(resultValue);
+        masm.and(64, array1Alignment, array1Address, 31);
+        masm.sub(64, array2Address, array2Address, array1Alignment);
+        masm.bic(64, array1Address, array1Address, 31);
```

Comment:
@SwapnilGaikwad I changed this to have array1 aligned after the first loop iteration. Can you check to see how this performs relative to having unaligned accesses the entire time?

Reply:
Sure